packages/components/src/tab-panel/index.tsx

```diff
@@ -92,113 +74,167 @@ const UnforwardedTabPanel = (
 	ref: ForwardedRef< any >
 ) => {
 	const instanceId = useInstanceId( TabPanel, 'tab-panel' );
-	const [ selected, setSelected ] = useState< string >();
 
-	const handleTabSelection = useCallback(
-		( tabKey: string ) => {
-			setSelected( tabKey );
-			onSelect?.( tabKey );
+	const prependInstanceId = useCallback(
+		( tabName: string | undefined ) => {
+			if ( typeof tabName === 'undefined' ) {
+				return;
+			}
+			return `${ instanceId }-${ tabName }`;
 		},
-		[ onSelect ]
+		[ instanceId ]
 	);
 
-	// Simulate a click on the newly focused tab, which causes the component
-	// to show the `tab-panel` associated with the clicked tab.
-	const activateTabAutomatically = (
-		_childIndex: number,
-		child: HTMLElement
-	) => {
-		child.click();
-	};
-	const selectedTab = tabs.find( ( { name } ) => name === selected );
-	const selectedId = `${ instanceId }-${ selectedTab?.name ?? 'none' }`;
+	// Separate the actual tab name from the instance ID. This is
+	// necessary because Ariakit internally uses the element ID when
+	// a new tab is selected, but our implementation looks specifically
+	// for the tab name to be passed to the `onSelect` callback.
+	const extractTabName = useCallback( ( id: string | undefined | null ) => {
+		if ( typeof id === 'undefined' || id === null ) {
+			return;
+		}
+		return id.match( /^tab-panel-[0-9]*-(.*)/ )?.[ 1 ];
+	}, [] );
```

Comment:
This is a pure function so we could probably just move it out of the component instead of doing a `useCallback`.

Reply:
👍 7b9a0c6c65